### ISSUE TITLE
Access violation in destructor of interpreted_path

### ISSUE THREAD

**Issue by mateka** [mikebmcl/P0267_RefImpl#69]
Is it ok for the following code to trigger "read access violation" in destructor of interpreted_path?
```
#include <io2d.h>

using namespace std;
using namespace std::experimental;
using namespace std::experimental::io2d;

int main() {
	auto tileset = image_surface{ filesystem::path{"tileset.png"}, image_file_format::png, format::argb32 };
	brush tilesetBrush{ std::move(tileset) };

	auto display = make_display_surface(640, 480, format::argb32, scaling::letterbox, refresh_rate::fixed, 60.0f);
	brush backBrush{ rgba_color{255.0f, 0.5f, 0.5f} };
	display.letterbox_brush(backBrush);
       
	auto bb = bounding_box{ 0.2f, 0.2f, 0.2f, 0.2f };
	auto cp = clip_props{ bb };
	display.draw_callback([&tilesetBrush, &cp](display_surface& srf) {
		srf.paint(tilesetBrush, nullopt, nullopt, cp);
	});
	return display.begin_show();
}
```

**Comment by mikebmcl** [comment]
It's never ok for destructors to throw and that's not the intention in this reference implementation. That said, the current design prototype which will be found in the R7 version of P0267 once it is published has a lot of changes. I have been meaning to make what was the master branch something tagged P0267R6, since that was what it implemented, and then make the draft work the master branch. I have just made those changes.

With the changes, your code repro needed a few tweaks. First, I used a png file included with sample_app. Then I changed the code as follows:

`int main() {
	auto tileset = image_surface{ filesystem::path{ "2016_06_22.png" }, image_file_format::png, format::argb32 };
	brush tilesetBrush{ std::move(tileset) };

	output_surface display(640, 480, format::argb32, scaling::letterbox, refresh_style::fixed, 60.0f);
	brush backBrush{ rgba_color{ 255.0f, 0.5f, 0.5f } };
	display.letterbox_brush(backBrush);

	auto bb = bounding_box{ 0.2f, 0.2f, 0.2f, 0.2f };
	auto cp = clip_props{ bb };
	display.draw_callback([&tilesetBrush, &cp](display_surface& srf) {
		srf.paint(tilesetBrush, nullopt, nullopt/*, cp*/);
	});
	return display.begin_show();
}
`

The factory function for `display_surface` is gone as is the `display_surface` class. In its place is the `basic_output_surface` class template, which has an alias of `output_surface` for its default instantiation.

Long story short, Running your code with the above modifications using the most up-to-date revisions of the reference implementation doesn't seem to produce errors on my end. My guess would be that the switch to using GraphicsMagick probably resolved the issue. I'll leave it open for now. If you can repro it again with the current master branch, please let me know so we can try to find the source of the problems.

Best regards,

-Mike

**Comment by mateka** [comment]
Thanks for a quick response! After upgrading master of io2d, clipping works! There are now two minor problems:
1. Warning:  p0267_refimpl\p0267_refimpl\xcairo_surfaces_image_impl.h(500): warning C4996: 'strncpy': This function or variable may be unsafe. Consider using strncpy_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.
2. Reading png files with alpha channel (32 bits). It looks like channels are swapped out. Reading 24 bit file from sample app works.

Best regards,
Mateusz

**Comment by mikebmcl** [comment]
1 is covered by the code as-is, but I'll try to change it to still do the right thing without triggering an MSVC warning.

2 is definitely a bug; I've repro'd it locally and am working to fix it. I'll post again when I've pushed a commit that I think resolves it.

Thanks!

-Mike
